common/appscale/common/service_stats/stats_manager.py

```diff
@@ -0,0 +1,492 @@
+    """ Provides value of metrics for recent requests which were finished
+    after specified timestamp (cursor).
+
+    Args:
+      cursor: a unix timestamp (in ms).
+      metrics_map: a dict describing metrics config.
+    Returns:
+      a dictionary containing value of metrics for recent requests.
+    """
+    requests = self._get_requests(since=cursor)
+    if not metrics_map:
+      metrics_map = self._metrics_for_recent_config
+    stats = self._render_recent(metrics_map, requests)
+    if not requests:
+      now = _now()
+      stats["from"] = now
+      stats["to"] = now
+    else:
+      stats["from"] = requests[0].end_time
+      stats["to"] = requests[-1].end_time
+    return stats
+
+  def _render_recent(self, metrics_config, requests):
+    """ Computes configured metrics according to metrics_config for requests.
+
+    Args:
+      metrics_config: a dictionary describing what metrics should be computed.
+      requests: a list of requests to compute metrics for.
+    Returns:
+      a dictionary containing computed metrics.
+    """
+    stats_dict = {}
+    for metric_pair in iteritems(metrics_config):
+      # Metrics config can contain following types of items:
+      #  - str->Metric
+      #  - Categorizer->Metric
+      #  - Categorizer->nested config with the same structure
+
+      if isinstance(metric_pair[0], str):
+        # Compute single metric if key is str
+        metric_name = metric_pair[0]
+        metric = metric_pair[1]
+        stats_dict[metric_name] = metric.compute(requests)
+        continue
+
+      # metric_pair[0] is instance of categorizers.Categorizer
+      categorizer = metric_pair[0]
+      stats_dict[categorizer.name] = categories_stats = {}
+
+      # Grouping requests by category
+      grouped_by_category = defaultdict(list)
+      for request_info in requests:
+        category = categorizer.category_of(request_info)
+        grouped_by_category[category].append(request_info)
+
+      if isinstance(metric_pair[1], metrics.Metric):
+        # Compute single metric for each category
+        metric = metric_pair[1]
+        for category, requests_group in iteritems(grouped_by_category):
+          categories_stats[category] = metric.compute(requests_group)
+      else:
+        # Render nested stats for each category
+        nested_config = metric_pair[1]
+        for category, requests_group in iteritems(grouped_by_category):
+          categories_stats[category] = self._render_recent(
+            nested_config, requests_group
+          )
+    return stats_dict
+
+  def _get_requests(self, since=None):
+    """ Selects requests which were finished since specified timestamp (in ms).
+
+    Args:
+      since: a unix timestamp in ms.
+    Returns:
+      a list of requests finished since specified timestamp.
+    """
+    if since is None:
+      return self._finished_requests
+    # Find the first element newer than 'since' using bisect
+    left, right = 0, len(self._finished_requests)
```

Comment:
Just a side note, if you end up using a deque for requests, it would be more efficient to do a linear scan than to bisect here.